mbuild/port.py

```diff
@@ -45,19 +55,32 @@ def __init__(self, anchor=None, orientation=None, separation=0):
         down = clone(up)
         down.rotate(np.pi, [0, 0, 1])
 
-        self.add(up, 'up')
-        self.add(down, 'down')
+        '''
+        Ports along antiparallel vectors should have 'up' and 'down' subports
+        at equivalent positions, except with reverse labels (i.e. the 'up'
+        subport for a port oriented in one direction will have the same xyz
+        positions as the 'down' port oriented in the opposite direction. As
+        a bit of a trick, when the z coordinate of the orientation vector is
+        negative the subports are created as if the vector was of the opposite
+        orientation and the labels are switched manually.
+        '''
+        swap_labels = False
+        if orientation[2] < 0:
```

Comment:
This seems really strange and if at all possible I would like to understand (and fix) the root of the problem instead of circumventing it. Why is this specific to the z-dimension? Does this have to do with the rotation above? `down.rotate(np.pi, [0, 0, 1])`

Reply:
Yes it is definitely strange, and I agree we should probably figure out the actual problem. I don't think it has to do with `down.rotate(np.pi, [0, 0, 1])`, since all this is doing is orienting the `down` subport in the opposite direction of `up`, prior to aligning along the `orientation` vector. But yes, it is very strange the problem is only specific to the -_z_ dimension. I've spent a decent amount of time trying to figure this out and have still not quite been able to wrap my head around it, but I will try to attach a Jupyter notebook sometime this weekend that details and visualizes the issue.

Reply:
I haven't tested this yet, but after some discussions yesterday and last night, I think the issue could be related to the z-direction due to the way we are calculating the cross product. We always calculate it in relation to the default direction and the orientation. We then always rotate using the `rotate` method. There must be some consistent method in which we are doing this rotation that makes the z-direction the only direction that must be flipped.